### PR TITLE
ddl: not to overwrite column charsets when altering table charset without 'convert to'

### LINKED ISSUE #14284: ddl: incompatible behavior with MySQL in `ALTER TABLE ... CHARSET` syntax
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
```
mysql> create table t(a varchar(5) charset utf8) charset utf8;
mysql> alter table t charset utf8mb4;
mysql> show create table t;
```

2. What did you expect to see?
In MySQL:
```
+-------+---------------------------------------------------+
| Table | Create Table                                      |
+-------+---------------------------------------------------+
| t     | CREATE TABLE `t` (                                |
|       |   `a` varchar(20) CHARACTER SET utf8 DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4           |
+-------+---------------------------------------------------+
```
3. What did you see instead?
```
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` varchar(20) DEFAULT NULL                              |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

Notice that `ALTER TABLE ... CHARSET` only changes the default charset of a table, the charset of existing columns should not be modified(`ALTER TABLE ... CONVERT TO CHARSET` should do the change).
 
4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
This happens to all TiDB versions.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve? 
Fix #14284.

PTAL at the related parser PR: pingcap/parser#706.

### What is changed and how it works?

- Modify `getCharsetAndCollateInTableOption`.
- Add an argument `needOverwriteCols` in DDL job of the type `ActionModifyTableCharsetAndCollate`.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

Release note

 - ddl: not to overwrite column charsets when altering table charset without 'convert to'

### CODE REVIEW

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -743,6 +743,31 @@ func (s *testIntegrationSuite4) TestChangingTableCharset(c *C) {
...
+	tk.MustExec("create table t(a varchar(5) charset utf8 collate utf8_unicode_ci) charset utf8 collate utf8_unicode_ci")
+	tk.MustExec("alter table t collate utf8_danish_ci")
+	tbl = testGetTableByName(c, s.ctx, "test", "t")
+	c.Assert(tbl, NotNil)
+	c.Assert(tbl.Meta().Charset, Equals, "utf8")
+	c.Assert(tbl.Meta().Collate, Equals, "utf8_danish_ci")
+	for _, col := range tbl.Meta().Columns {
+		c.Assert(col.Charset, Equals, "utf8")
+		// Column collate should remain unchanged.
+		c.Assert(col.Collate, Equals, "utf8_unicode_ci")
```
Comment (AilinKid) [inline ddl/db_integration_test.go]: Maybe you should add a case for `convert to`—— both table and column's charset are changed.

Comment (tangenta) [reply ddl/db_integration_test.go]: The case already exists:
https://github.com/pingcap/tidb/blob/d5be4b6c485ec80cb5acd69dcca71feefd33433b/ddl/db_integration_test.go#L657-L671

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -3091,7 +3093,7 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 		SchemaName: schema.Name.L,
 		Type:       model.ActionModifyTableCharsetAndCollate,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{toCharset, toCollate},
+		Args:       []interface{}{toCharset, toCollate, needsOverwriteCols},
```
Comment (wjhuang2016) [inline ddl/ddl_api.go]: Will this cause compatible problem? @zimulala PTAL

Comment (zimulala) [reply ddl/ddl_api.go]: I think it is fine. When rolling upgrade, it is equivalent to not solving this problem.

File: ddl/db_integration_test.go (hunk 1 above)
Comment (AilinKid) [reply ddl/db_integration_test.go]: got

**Comment by reafans** [comment]
LGTM

**Comment by reafans** [comment]
lgtm

**Pull request merged**